Introduce `suppress` to streams.

truemped opened the pull request:
Version 2.1 introduced a new operator: [suppress](https://kafka.apache.org/21/documentation/streams/developer-guide/dsl-api.html#window-final-results). It allows to only
forward the final result of a windowed aggregation.

This PR adds the `suppress` function to the streams namespace and
allows for the new operator to be used in Kafka Streams apps. It works but may not be in a final stage. Asking for feedback at this point whether this is what you have in mind or I should change something.

Best
Daniel

99-not-out commented:
Hey - thanks for this Daniel.
We will do a review soon and update here.

Regards
Matt

cddr commented on src/jackdaw/streams/interop.clj:
```diff
@@ -389,6 +400,15 @@
+    [_ {:keys [max-records max-bytes until-time-limit-ms]
+        :as suppress}]
+    (let [cfg (suppress-config suppress)]
+      (println cfg)
```
Shall we lose the `println` which I presume was there for debugging purposes.

truemped replied:
ups...

cddr commented on src/jackdaw/test.clj:
```diff
@@ -67,7 +67,7 @@
 (defn consume-all-outputs!
```
We replaced this test namespace with [the test machine](https://github.com/FundingCircle/jackdaw/pull/28) (apologies for the breaking change). There will be a conflict on this file when you update the master branch. You can resolve by simply accepting the upstream branch.

truemped replied:
+1

cddr commented on test/jackdaw/streams_test.clj:
```diff
@@ -603,7 +604,45 @@
-            (is (= [1 6] (nth keyvals 2)))))))))
+            (is (= [1 6] (nth keyvals 2))))))))
+
+  (testing "suppress records"
```
Shall we add another example that covers the use of a suppress config (e.g. max-records, until-time-limit etc).

truemped replied:
Will do. I'll try to get to it by next Monday!

truemped commented:
Hey, @cddr. It's been a while. I've updated the PR and added a lot more test cases. Let me know what you think. Would be great to get a new version with including this out, once it's in a state that works for  you.

Best
Daniel

cddr commented on .lein-failures:
```diff
@@ -0,0 +1 @@
+{"jackdaw.streams-test" ["KTable"]}
```
Suspect this was not intended to be checked in?

truemped replied:
No, removed

cddr commented on src/jackdaw/streams/interop.clj:
```diff
@@ -396,6 +408,14 @@
                  ^ValueJoiner (value-joiner value-joiner-fn))))
 
+  (suppress
+    [_ {:keys [max-records max-bytes until-time-limit-ms]
```
Seems like you no longer need to destructure the `suppress` argument here.

truemped replied:
Yes, updated the code and also removed the `let` binding. No need for it

truemped commented:
Cool, thanks @cddr. Anything I still should do/help do to get it merged? Some circleci job fails, but that does not seem to be test related.

vijumathew commented on src/jackdaw/streams/protocols.clj:
```diff
@@ -192,6 +192,14 @@
+    be unbounded. If `until-time-limit-ms` is set, this will override the
+    `TimeWindow` interval. Note that when relying on the configured `TimeWindow`
+    the default `grace` period is `24h - window-size`.")
+
```
could be helpful to format the docstring after the styleguide. just ending the first line after the first sentence and starting "You can either specify..." on a newline

https://github.com/bbatsov/clojure-style-guide#docstring-summary

truemped replied:
Updated

vijumathew commented on src/jackdaw/streams/interop.clj:
```diff
@@ -396,6 +408,11 @@
+  (suppress
+    [_ suppress]
+    (clj-ktable
+       (.suppress ktable (suppress-config suppress))))
```
it looks like the spec defined before was named `suppress-config` but the fn that uses it is called `suppress-config` and the config is named `suppress`. this seems a little confusing to me. 

might be helpful to name the config arg `suppress-config` to match the spec and the fn `create-suppress-config` or something. the names i suggested could definitely be better though

truemped replied:
I've tried to make more sense of the names. The argument is now `suppress-config` and the function is `suppress-config->suppressed` which is what it's doing actually: turning a `suppress-config` into a `org.apache.kafka.streams.kstream.Suppressed`.

vijumathew commented on test/jackdaw/streams_test.clj:
```diff
@@ -643,7 +644,177 @@
+          topic-c (mock/topic "topic-c")
+
+          ;; if we don't set the `grace` period of the `TimeWindows`, the
+          ;; default is used: 24h - window
```
including this comment once seems sufficient to me. could get rid of this

truemped replied:
removed

truemped commented:
Hey @vijumathew, could you have another look at this?
